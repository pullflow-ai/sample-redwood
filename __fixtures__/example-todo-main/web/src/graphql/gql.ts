/* eslint-disable */
import * as types from "./graphql";
import { TypedDocumentNode as DocumentNode } from "@graphql-typed-document-node/core";

/**
 * Map of all GraphQL operations in the project.
 *
 * This map has several performance disadvantages:
 * 1. It is not tree-shakeable, so it will include all operations in the project.
 * 2. It is not minifiable, so the string of a GraphQL query will be multiple times inside the bundle.
 * 3. It does not support dead code elimination, so it will add unused operations.
 *
 * Therefore it is highly recommended to use the babel or swc plugin for production.
 */
const documents = {
  "\n  mutation AddTodo_CreateTodo($body: String!) {\n    createTodo(body: $body) {\n      id\n      __typename\n      body\n      status\n    }\n  }\n":
    types.AddTodo_CreateTodoDocument,
  "\n  query NumTodosCell_GetCount {\n    todosCount\n  }\n":
    types.NumTodosCell_GetCountDocument,
  "\n  query TodoListCell_GetTodos {\n    todos {\n      id\n      body\n      status\n    }\n  }\n":
    types.TodoListCell_GetTodosDocument,
  "\n  mutation TodoListCell_CheckTodo($id: Int!, $status: String!) {\n    updateTodoStatus(id: $id, status: $status) {\n      id\n      __typename\n      status\n    }\n  }\n":
    types.TodoListCell_CheckTodoDocument,
};

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 *
 *
 * @example
 * ```ts
 * const query = graphql(`query GetUser($id: ID!) { user(id: $id) { name } }`);
 * ```
 *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(
  source: "\n  mutation AddTodo_CreateTodo($body: String!) {\n    createTodo(body: $body) {\n      id\n      __typename\n      body\n      status\n    }\n  }\n"
): (typeof documents)["\n  mutation AddTodo_CreateTodo($body: String!) {\n    createTodo(body: $body) {\n      id\n      __typename\n      body\n      status\n    }\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(
  source: "\n  query NumTodosCell_GetCount {\n    todosCount\n  }\n"
): (typeof documents)["\n  query NumTodosCell_GetCount {\n    todosCount\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(
  source: "\n  query TodoListCell_GetTodos {\n    todos {\n      id\n      body\n      status\n    }\n  }\n"
): (typeof documents)["\n  query TodoListCell_GetTodos {\n    todos {\n      id\n      body\n      status\n    }\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(
  source: "\n  mutation TodoListCell_CheckTodo($id: Int!, $status: String!) {\n    updateTodoStatus(id: $id, status: $status) {\n      id\n      __typename\n      status\n    }\n  }\n"
): (typeof documents)["\n  mutation TodoListCell_CheckTodo($id: Int!, $status: String!) {\n    updateTodoStatus(id: $id, status: $status) {\n      id\n      __typename\n      status\n    }\n  }\n"];

export function graphql(source: string) {
  return (documents as any)[source] ?? {};
}

export type DocumentType<TDocumentNode extends DocumentNode<any, any>> =
  TDocumentNode extends DocumentNode<infer TType, any> ? TType : never;

export function gql(source: string | TemplateStringsArray) {
  if (typeof source === "string") {
    return graphql(source);
  }

  return graphql(source.join("\n"));
}
