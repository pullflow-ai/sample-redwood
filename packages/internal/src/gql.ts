import { CodeFileLoader } from '@graphql-tools/code-file-loader'
import { loadSchema } from '@graphql-tools/load'
import type {
  DocumentNode,
  FieldNode,
  InlineFragmentNode,
  OperationDefinitionNode,
  OperationTypeNode,
} from 'graphql'
import { Kind, parse, print, visit } from 'graphql'

import { rootSchema } from '@redwoodjs/graphql-server'
import { getPaths } from '@redwoodjs/project-config'

interface Operation {
  operation: OperationTypeNode
  name: string | undefined
  fields: (string | Field)[]
}

interface Field {
  string: (string | Field)[]
}

export const parseGqlQueryToAst = (gqlQuery: string) => {
  const ast = parse(gqlQuery)
  return parseDocumentAST(ast)
}

export const parseDocumentAST = (document: DocumentNode) => {
  const operations: Operation[] = []

  visit(document, {
    OperationDefinition(node: OperationDefinitionNode) {
      const fields: any[] = []

      node.selectionSet.selections.forEach((field) => {
        fields.push(getFields(field as FieldNode))
      })

      operations.push({
        operation: node.operation,
        name: node.name?.value,
        fields,
      })
    },
  })

  return operations
}

const getFields = (field: FieldNode): any => {
  // base
  if (!field.selectionSet) {
    return field.name.value
  } else {
    const obj: Record<string, FieldNode[]> = {
      [field.name.value]: [],
    }

    const lookAtFieldNode = (node: FieldNode | InlineFragmentNode): void => {
      node.selectionSet?.selections.forEach((subField) => {
        switch (subField.kind) {
          case Kind.FIELD:
            obj[field.name.value].push(getFields(subField))
            break
          case Kind.FRAGMENT_SPREAD:
            // TODO: Maybe this will also be needed, right now it's accounted for to not crash in the tests
            break
          case Kind.INLINE_FRAGMENT:
            lookAtFieldNode(subField)
        }
      })
    }

    lookAtFieldNode(field)

    return obj
  }
}

export const listQueryTypeFieldsInProject = async () => {
  try {
    const schemaPointerMap = {
      [print(rootSchema.schema)]: {},
      'graphql/**/*.sdl.{js,ts}': {},
      'directives/**/*.{js,ts}': {},
      'subscriptions/**/*.{js,ts}': {},
    }

    const mergedSchema = await loadSchema(schemaPointerMap, {
      loaders: [
        new CodeFileLoader({
          noRequire: true,
          pluckConfig: {
            globalGqlIdentifierName: 'gql',
          },
        }),
      ],
      cwd: getPaths().api.src,
      assumeValidSDL: true,
    })

    const queryTypeFields = mergedSchema.getQueryType()?.getFields()

    // Return empty array if no schema found
    return Object.keys(queryTypeFields ?? {})
  } catch (e) {
    console.error(e)
    return []
  }
}
