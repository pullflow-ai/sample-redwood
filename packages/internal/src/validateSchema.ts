import { CodeFileLoader } from '@graphql-tools/code-file-loader'
import { loadTypedefs } from '@graphql-tools/load'
import { mergeTypeDefs } from '@graphql-tools/merge'
import type { DocumentNode, ObjectTypeDefinitionNode } from 'graphql'
import { Kind, visit } from 'graphql'

import { rootSchema } from '@redwoodjs/graphql-server'
import { getPaths } from '@redwoodjs/project-config'

import { isServerFileSetup, isRealtimeSetup } from './project'

export const DIRECTIVE_REQUIRED_ERROR_MESSAGE =
  'You must specify one of @requireAuth, @skipAuth or a custom directive'

export const DIRECTIVE_INVALID_ROLE_TYPES_ERROR_MESSAGE =
  'Please check that the requireAuth roles is a string or an array of strings.'

/**
 * These are names that are commonly used in GraphQL schemas as scalars
 * and would cause a conflict if used as a type name.
 *
 * Note: Query, Mutation, and Subscription are not included here because
 * they are checked for separately.
 */
export const RESERVED_TYPES = [
  'Int',
  'Float',
  'Boolean',
  'String',
  'DateTime',
  'ID',
  'uid',
  'as',
]

export function validateSchema(
  schemaDocumentNode: DocumentNode,
  typesToCheck: string[] = ['Query', 'Mutation'],
) {
  const validationOutput: string[] = []
  const reservedNameValidationOutput: Record<string, any> = []
  const directiveRoleValidationOutput: Record<string, any> = []

  // Is Subscriptions are enabled with Redwood Realtime, then enforce a rule
  // that a Subscription type needs to have a authentication directive applied,
  // just as Query and Mutation requires
  if (isServerFileSetup() && isRealtimeSetup()) {
    typesToCheck.push('Subscription')
  }

  visit(schemaDocumentNode, {
    InterfaceTypeDefinition(typeNode) {
      // Warn that an interface definition in the SDL is using a reserved GraphQL type
      if (RESERVED_TYPES.includes(typeNode.name.value)) {
        reservedNameValidationOutput.push({
          objectType: 'interface',
          name: typeNode.name.value,
        })
      }
    },
    InputObjectTypeDefinition(typeNode) {
      // Warn that an input definition in the SDL is using a reserved GraphQL type
      if (RESERVED_TYPES.includes(typeNode.name.value)) {
        reservedNameValidationOutput.push({
          objectType: 'input type',
          name: typeNode.name.value,
        })
      }
    },
    ObjectTypeDefinition(typeNode) {
      // Warn that a type definition in the SDL is using a reserved GraphQL type
      if (RESERVED_TYPES.includes(typeNode.name.value)) {
        reservedNameValidationOutput.push({
          objectType: 'type',
          name: typeNode.name.value,
        })
      }
      if (typesToCheck.includes(typeNode.name.value)) {
        for (const field of typeNode.fields ||
          ([] as ObjectTypeDefinitionNode[])) {
          const fieldName = field.name.value
          const fieldTypeName = typeNode.name.value

          const isRedwoodQuery =
            fieldName === 'redwood' && fieldTypeName === 'Query'
          const isCurrentUserQuery =
            fieldName === 'currentUser' && fieldTypeName === 'Query'
          // skip validation for redwood query and currentUser
          if (!(isRedwoodQuery || isCurrentUserQuery)) {
            const hasDirective = field.directives?.length

            if (!hasDirective) {
              validationOutput.push(`${fieldName} ${fieldTypeName}`)
            }

            // we want to check that the requireAuth directive roles argument value
            // is a string or an array of strings
            field.directives?.forEach((directive) => {
              if (directive.name.value === 'requireAuth') {
                directive.arguments?.forEach((arg) => {
                  if (arg.name.value === 'roles') {
                    if (
                      arg.value.kind !== Kind.STRING &&
                      arg.value.kind !== Kind.LIST
                    ) {
                      directiveRoleValidationOutput.push({
                        fieldName: fieldName,
                        invalid: arg.value.kind,
                      })
                    }

                    // check list (array)
                    if (arg.value.kind === Kind.LIST) {
                      const invalidValues = arg.value.values?.filter(
                        (val) => val.kind !== Kind.STRING,
                      )
                      if (invalidValues.length > 0) {
                        invalidValues.forEach((invalid) => {
                          directiveRoleValidationOutput.push({
                            fieldName: fieldName,
                            invalid: invalid.kind,
                          })
                        })
                      }
                    }
                  }
                })
              }
            })
          }
        }
      }
    },
  })

  if (validationOutput.length > 0) {
    const fieldsWithoutDirectives = validationOutput.map(
      (field) => `- ${field}`,
    )

    throw new Error(
      `${DIRECTIVE_REQUIRED_ERROR_MESSAGE} for\n${fieldsWithoutDirectives.join(
        '\n',
      )} \n`,
    )
  }

  if (directiveRoleValidationOutput.length > 0) {
    const fieldWithInvalidRoleValues = directiveRoleValidationOutput.map(
      (field: Record<string, any>) =>
        `- ${field.fieldName} has an invalid ${field.invalid}`,
    )

    throw new RangeError(
      `${DIRECTIVE_INVALID_ROLE_TYPES_ERROR_MESSAGE}\n\n${fieldWithInvalidRoleValues.join(
        '\n',
      )} \n\nFor example: @requireAuth(roles: "admin") or @requireAuth(roles: ["admin", "editor"])`,
    )
  }

  if (reservedNameValidationOutput.length > 0) {
    const reservedNameMsg = reservedNameValidationOutput.map(
      (output: Record<string, any>) => {
        return `The ${output.objectType} named '${output.name}' is a reserved GraphQL name.\nPlease rename it to something more specific, like: Application${output.name}.\n`
      },
    )
    throw new TypeError(reservedNameMsg.join('\n'))
  }
}

export const loadAndValidateSdls = async () => {
  const projectTypeSrc = await loadTypedefs(
    [
      'graphql/**/*.sdl.{js,ts}',
      'directives/**/*.{js,ts}',
      'subscriptions/**/*.{js,ts}',
    ],
    {
      loaders: [
        new CodeFileLoader({
          noRequire: true,
          pluckConfig: {
            globalGqlIdentifierName: 'gql',
          },
        }),
      ],
      cwd: getPaths().api.src,
    },
  )

  // The output of the above function doesn't give us the documents directly
  const projectDocumentNodes = Object.values(projectTypeSrc)
    .map(({ document }) => document)
    .filter((documentNode): documentNode is DocumentNode => {
      return !!documentNode
    })

  // Merge in the rootSchema with JSON scalars, etc.
  const mergedDocumentNode = mergeTypeDefs([
    rootSchema.schema,
    projectDocumentNodes,
  ])

  validateSchema(mergedDocumentNode)
}
