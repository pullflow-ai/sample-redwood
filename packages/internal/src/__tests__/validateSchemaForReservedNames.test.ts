import path from 'path'

import gql from 'graphql-tag'
import { beforeAll, afterAll, describe, test, expect } from 'vitest'

import { validateSchema } from '../validateSchema'

const FIXTURE_PATH = path.resolve(
  __dirname,
  '../../../../__fixtures__/example-todo-main',
)

beforeAll(() => {
  process.env.RWJS_CWD = FIXTURE_PATH
})
afterAll(() => {
  delete process.env.RWJS_CWD
})

describe('SDL with no reserved names used', () => {
  describe('SDL is valid', () => {
    test('with proper type definition names', () => {
      const document = gql`
        type Message {
          from: String
          body: String
        }

        type Query {
          room(id: ID!): [Message!]! @skipAuth
        }

        input SendMessageInput {
          roomId: ID!
          from: String!
          body: String!
        }

        type Mutation {
          sendMessage(input: SendMessageInput!): Message! @skipAuth
        }
      `

      expect(() => validateSchema(document)).not.toThrowError()
    })
    test('with proper interface interface definition names', () => {
      const document = gql`
        interface Node {
          id: ID!
        }

        type Message implements Node {
          id: ID!
          from: String
          body: String
        }

        type Query {
          room(id: ID!): [Message!]! @skipAuth
        }
      `
      expect(() => validateSchema(document)).not.toThrowError()
    })
    test('with proper interface input type definition names', () => {
      const document = gql`
        type Message {
          from: String
          body: String
        }

        type Query {
          room(id: ID!): [Message!]! @skipAuth
        }

        input SendMessageInput {
          roomId: ID!
          from: String!
          body: String!
        }

        type Mutation {
          sendMessage(input: SendMessageInput!): Message! @skipAuth
        }
      `
      expect(() => validateSchema(document)).not.toThrowError()
    })
  })

  describe('SDL is invalid', () => {
    test('because uses a reserved name as a type', () => {
      const document = gql`
        type Float {
          from: String
          body: String
        }

        type Query {
          room(id: ID!): [Message!]! @skipAuth
        }

        input SendMessageInput {
          roomId: ID!
          from: String!
          body: String!
        }

        type Mutation {
          sendMessage(input: SendMessageInput!): Message! @skipAuth
        }
      `
      expect(() => validateSchema(document)).toThrowErrorMatchingSnapshot()
    })
  })

  test('because uses a reserved name as an input', () => {
    const document = gql`
      type Message {
        from: String
        body: String
      }

      type Query {
        room(id: ID!): [Message!]! @skipAuth
      }

      input Float {
        roomId: ID!
        from: String!
        body: String!
      }

      type Mutation {
        sendMessage(input: SendMessageInput!): Message! @skipAuth
      }
    `
    expect(() => validateSchema(document)).toThrowErrorMatchingSnapshot()
  })

  test('because uses a reserved name as an interface', () => {
    const document = gql`
      interface Float {
        id: ID!
      }

      type Message implements Float {
        id: ID!
        from: String
        body: String
      }

      type Query {
        room(id: ID!): [Message!]! @skipAuth
      }

      input SendMessageInput {
        roomId: ID!
        from: String!
        body: String!
      }

      type Mutation {
        sendMessage(input: SendMessageInput!): Message! @skipAuth
      }
    `
    expect(() => validateSchema(document)).toThrowErrorMatchingSnapshot()
  })
})
