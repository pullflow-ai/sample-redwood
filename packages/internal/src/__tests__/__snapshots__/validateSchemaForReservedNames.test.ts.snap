// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SDL with no reserved names used > SDL is invalid > because uses a reserved name as a type 1`] = `
[TypeError: The type named 'Float' is a reserved GraphQL name.
Please rename it to something more specific, like: ApplicationFloat.
]
`;

exports[`SDL with no reserved names used > because uses a reserved name as an input 1`] = `
[TypeError: The input type named 'Float' is a reserved GraphQL name.
Please rename it to something more specific, like: ApplicationFloat.
]
`;

exports[`SDL with no reserved names used > because uses a reserved name as an interface 1`] = `
[TypeError: The interface named 'Float' is a reserved GraphQL name.
Please rename it to something more specific, like: ApplicationFloat.
]
`;
